### PR TITLE
load following videos, add button login

### PR DESCRIPTION
**Pull request by bogdanStepanyuk**

### CODE REVIEW

**Comment by AndriyBas** [comment]
@bogdanStepanyuk please merge with `develop`

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 1]
```diff
@@ -97,6 +105,14 @@ public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
         });
     }
 
+    protected void updateLoginUI() {}
+
+    @OnClick((R.id.loginIn))
+    public void loginUp() {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk should be private

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 2]
```diff
@@ -32,14 +35,15 @@
...
-    protected static String query = "";
 
     @Inject
     protected ApiService apiService;
     @Inject
     protected Gson gson;
     @Inject
     protected AppPreferences prefs;
+    @Bind(R.id.loginIn)
+    Button loginIn;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk 
1) this is very bad name for variable. Please change it to smth meaningful
2) Also, please use access modifiers. 
3) also, please update button id in XML (use snake case, add scope as a prefix)

File: app/src/main/java/com/uptech/example/vidme/ui/front/MainActivity.java [hunk 3]
```diff
@@ -93,6 +95,12 @@ public boolean onOptionsItemSelected(MenuItem item) {
                  return true;
              case R.id.front_page_action_log_out:
                  prefs.logOut();
+
+                 dataChangeListener = (DataChangeListener) sectionsPagerAdapter.getItem(2);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/MainActivity.java]: @bogdanStepanyuk 
1) we cannot use constant `2` here. 
2) `dataChangeListener` should be initialised before, not here

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java [hunk 4]
```diff
@@ -113,6 +118,13 @@ public void onRefresh() {
...
+
+        if (activity instanceof DataChangeListener) {
+            dataChangeListener = (DataChangeListener) getActivity();
+        } else {
+            throw new RuntimeException("activity is not DataChangeListener");
+        }
+
         if (activity instanceof SearchQueryObservable) {
             ((SearchQueryObservable) activity).addListener(this);
         }
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java]: @bogdanStepanyuk throw exception here as well

**Comment by bogdanStepanyuk** [comment]
found some bags, now fixing this

File: app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java [hunk 5]
```diff
@@ -1,35 +1,93 @@
...
 /**
  * Created by oleh on 30.03.16
  */
-public class FollowingTabFragment extends BaseVideoTabFragment {
-
+public class FollowingTabFragment extends BaseVideoTabFragment implements DataChangeListener {
     /**
      * @param offset count of loaded videos
      */
+    protected boolean isLoginUpdate;
```
Comment (Dima564) [inline app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java]: What does 'isLoginUpdate' mean?

File: app/src/main/java/com/uptech/example/vidme/adapters/UsersSearchListAdapter.java [hunk 6]
```diff
@@ -31,6 +31,10 @@ public UsersSearchListAdapter(List<User> userList, Context context) {
         this.context = context;
     }
 
+    public void remuveUserList() {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/adapters/UsersSearchListAdapter.java]: @bogdanStepanyuk please fix typo

File: app/src/main/java/com/uptech/example/vidme/adapters/VideoListAdapter.java [hunk 7]
```diff
@@ -105,7 +107,9 @@ public Video getItem(int position) {
 
         @OnClick(R.id.video_item_layout)
         public void onItemClicked() {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/adapters/VideoListAdapter.java]: @bogdanStepanyuk plz change access modifier to `protected`

File: app/src/main/java/com/uptech/example/vidme/adapters/VideoSearchListAdapter.java [hunk 8]
```diff
@@ -32,6 +32,10 @@ public VideoSearchListAdapter(List<Video> videoList, Context context) {
         this.context = context;
     }
 
+    public void removeVideoList() {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/adapters/VideoSearchListAdapter.java]: @bogdanStepanyuk the name of the method is not correct, please rename it to `clearList` or `clearAdapter` or `clearListItems` or smth like that. Also rename similar method in User adapter

File: app/src/main/java/com/uptech/example/vidme/ui/BaseActivity.java [hunk 9]
```diff
@@ -26,6 +26,7 @@
 public abstract class BaseActivity extends AppCompatActivity {
 
     protected final String TAG = getClass().getSimpleName();
+    protected static final int RESULT_ACTIVITY_CODE = 911;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/BaseActivity.java]: @bogdanStepanyuk why is it here ? it should not be here

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 10]
```diff
@@ -17,29 +19,33 @@
...
+    protected static final int RESULT_ACTIVITY_CODE = 911;
 
     @Inject
     protected ApiService apiService;
     @Inject
     protected Gson gson;
     @Inject
     protected AppPreferences prefs;
+    @Bind(R.id.login_sign_in_button)
+    protected Button loginSignIn;
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk please rename button, as discussed

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 11]
```diff
@@ -97,6 +107,15 @@ public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
         });
     }
 
+    protected void updateLoginUI() {
+    }
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk empty method ?

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 12]
```diff
@@ -97,6 +107,15 @@ public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
...
+
+    @OnClick((R.id.login_sign_in_button))
+    protected void loginSignInButton() {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk please rename method

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 13]
```diff
@@ -117,4 +136,10 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
             }
         }
     }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        Log.d("des", "destroy");
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk do we need this ?

File: app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java [hunk 14]
```diff
@@ -28,8 +32,44 @@ protected void loadVideos(int offset) {
...
+    @Override
+    protected void updateLoginUI() {
+        super.updateLoginUI();
+        upDateUI(true);
+    }
+
+    @Override
+    public void updateData() {
+        Log.d("start", "updateData");
+        if (videoList != null && loginSignIn != null) {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java]: @bogdanStepanyuk plz fix variable names

File: app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java [hunk 15]
```diff
@@ -28,8 +32,44 @@ protected void loadVideos(int offset) {
...
+            upDateUI(true);
+        }
+    }
+
+    private void upDateUI(boolean update) {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java]: @bogdanStepanyuk please fix method name

Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java]: @bogdanStepanyuk argument name is not OK

File: app/src/main/java/com/uptech/example/vidme/ui/front/MainActivity.java [hunk 16]
```diff
@@ -76,42 +80,59 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
     public void startSearchActivity() {
         Intent i = new Intent(this, SearchActivity.class);
         startActivity(i);
     }
 
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+
+        if (resultCode == RESULT_ACTIVITY_CODE) {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/MainActivity.java]: @bogdanStepanyuk we need to check for `result ok`

File: app/src/main/java/com/uptech/example/vidme/ui/login/LoginActivity.java [hunk 17]
```diff
@@ -128,6 +127,10 @@ private void attemptLogin() {
                     .subscribe(userResponse -> {
                         showProgress(false);
                         prefs.saveToken(userResponse.getAuth().getToken());
+
+                        Intent intent = new Intent();
+                        setResult(RESULT_ACTIVITY_CODE, intent);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/login/LoginActivity.java]: @bogdanStepanyuk what us it ?

File: app/src/main/java/com/uptech/example/vidme/ui/login/SignUpActivity.java [hunk 18]
```diff
@@ -97,9 +96,8 @@ private void attemptSignUp() {
                     .subscribe(userResponse -> {
                         showProgress(false);
                         prefs.saveToken(userResponse.getAuth().getToken());
-                        Intent i = new Intent(SignUpActivity.this, MainActivity.class);
-                        i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
-                        startActivity(i);
+                        Intent intent = new Intent();
+                        setResult(RESULT_ACTIVITY_CODE, intent);
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/login/SignUpActivity.java]: @bogdanStepanyuk what is it ?

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java [hunk 19]
```diff
@@ -1,48 +1,58 @@
...
-                    }, error -> {
-                        onLoadingFailure();
-                    });
+    public String getTitle() {
+        if (getActivity() == null) {
+            return "";
+        } else if (adapter == null || adapter.getItemCount() == 0) {
+            return getString(R.string.users);
+        } else {
+            return getString(R.string.users) + " (" + total + ")";
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/UsersFragment.java]: @bogdanStepanyuk please use arguments in string resources as discussed

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java [hunk 20]
```diff
@@ -26,17 +27,17 @@
 public abstract class BaseSearchFragment extends BaseFragment implements SwipeRefreshLayout.OnRefreshListener
         , SearchActivity.SearchQueryListener {
 
-    private String query = "";
-    protected boolean shouldReplaceData;
+    protected String total = "";
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/BaseSearchFragment.java]: @bogdanStepanyuk I think it's smth wrong here . Why total is String ?

File: app/src/main/res/drawable/custom_button_login.xml [hunk 21]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+        <solid android:color="@color/background_button_login"/>
+        <corners android:radius="7dp"/>
```
Comment (AndriyBas) [inline app/src/main/res/drawable/custom_button_login.xml]: @bogdanStepanyuk plz use dimen res for this value

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 22]
```diff
@@ -111,9 +129,15 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode, resultCode, data);
         // if the result came from VideoActivity, we should update the video played
         // because the number of likes or comments could have changed
-        if (requestCode == VIDEO_PLAYER_REQUEST) {
-            if (resultCode == Activity.RESULT_OK) {
-                adapter.updateItem((Video) data.getSerializableExtra(VideoActivity.VIDEO_KEY));
+        if (resultCode == getActivity().RESULT_OK) {
+            if (requestCode == REQUEST_MY_SUPER_CODE) {
+                if (getActivity() instanceof MainActivity && getActivity() != null) {
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk `getActivity() != null` is a redundant check here

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 23]
```diff
@@ -111,9 +129,15 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode, resultCode, data);
         // if the result came from VideoActivity, we should update the video played
         // because the number of likes or comments could have changed
-        if (requestCode == VIDEO_PLAYER_REQUEST) {
-            if (resultCode == Activity.RESULT_OK) {
-                adapter.updateItem((Video) data.getSerializableExtra(VideoActivity.VIDEO_KEY));
+        if (resultCode == getActivity().RESULT_OK) {
+            if (requestCode == REQUEST_MY_SUPER_CODE) {
+                if (getActivity() != null) {
+                    ((MainActivity) getActivity()).updateViewPager();
```
Comment (AndriyBas) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: @bogdanStepanyuk plz refactor as discussed

File: app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java [hunk 24]
```diff
@@ -116,13 +133,37 @@ public void onItemClicked(Video video, int position) {
...
+
+    @Override
     public void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode, resultCode, data);
         // if the result came from VideoActivity, we should update the video played
         // because the number of likes or comments could have changed
-        if (requestCode == VIDEO_PLAYER_REQUEST) {
-            if (resultCode == Activity.RESULT_OK) {
-                adapter.updateItem((Video) data.getSerializableExtra(VideoActivity.VIDEO_KEY));
+        // TODO: 27.04.16 change with interface
```
Comment (Dima564) [inline app/src/main/java/com/uptech/example/vidme/ui/front/BaseVideoTabFragment.java]: Remove todo

File: app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java [hunk 25]
```diff
@@ -28,8 +32,44 @@ protected void loadVideos(int offset) {
...
+    public void updateData() {
+        Log.d("start", "updateData");
+        updateUI(true);
+    }
+
+    private void updateUI(boolean changeToList) {
+        if (videoList != null && buttonSignIn != null) {
+            if (changeToList) {
+                videoList.setVisibility(View.GONE);
+                buttonSignIn.setVisibility(View.VISIBLE);
```
Comment (Dima564) [inline app/src/main/java/com/uptech/example/vidme/ui/front/FollowingTabFragment.java]: Use ternary operator here

File: app/src/main/java/com/uptech/example/vidme/ui/front/MainActivity.java [hunk 26]
```diff
@@ -105,14 +115,38 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+
+        if (resultCode == RESULT_OK) {
+            if (requestCode == REQUEST_LOGGED_CODE) {
+                Log.d("code", resultCode + "");
+                updateViewPager();
+            } else {
+                Log.d("code", "0");
```
Comment (Dima564) [inline app/src/main/java/com/uptech/example/vidme/ui/front/MainActivity.java]: What is this?

File: app/src/main/java/com/uptech/example/vidme/ui/search/fragments/VideosFragment.java [hunk 27]
```diff
@@ -1,48 +1,60 @@
...
 
     @Override
-    public void addListener(PageTotalVideosListener pageTotalVideos) {
-        totalVideosListeners.add(pageTotalVideos);
+    public void loadData(int offset, String query) {
+        apiService.getSearchVideo(offset, query)
+                .subscribe(videoListResponse -> {
+                    if (offset == 0) {
+                        adapter.clearVideoList();
+                        adapter.notifyDataSetChanged();
```
Comment (Dima564) [inline app/src/main/java/com/uptech/example/vidme/ui/search/fragments/VideosFragment.java]: Do we have to notify dataset changed here?

File: app/src/main/java/com/uptech/example/vidme/ui/video_player/VideoActivity.java [hunk 28]
```diff
@@ -211,11 +213,41 @@ private void initLikeListener() {
...
+                    // needed to revert the likes count when this like was not cast
+                    videoModel.setScore(videoModel.getScore() - voteValue);
+                    videoToolbar.setVideoModel(videoModel);
+                });
                         // send a request to server
-                .flatMap(click -> videoModel.isLiked() ?
-                        likedVideosManager.likeVideo(videoModel, videoView.getCurrentPosition())
-                        : likedVideosManager.unlikeVideo(videoModel, videoView.getCurrentPosition()))
-                .subscribe(getLikeRequestSubscriber());
+//                .flatMap(click -> videoModel.isLiked() ?
```
Comment (Dima564) [inline app/src/main/java/com/uptech/example/vidme/ui/video_player/VideoActivity.java]: remove unused code

**Pull request merged**